torch/_dynamo/variables/user_defined.py

```diff
@@ -401,6 +401,15 @@ def var_getattr(self, tx, name):
                 subobj.__func__, self, source=source, **options
             )
         elif isinstance(subobj, types.FunctionType):
+            # Check `__dict__` to bypass the function descriptor protocol to
+            # accurately check for static method
+            is_staticmethod = name in type(self.value).__dict__ and isinstance(
```

Comment:
If people know of a better way to check this, feel free to let me know.

Reply:
maybe try something like this?
https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/nn_module.py#L222

Reply:
This does not seem to be correct :/

```
>>> import torch
>>> from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
>>> from torch._dynamo.utils import istype
>>> istype(FSDP.summon_full_params, staticmethod)
False
```
https://github.com/pytorch/pytorch/blob/5f138a6b6561cd6fa4653d06623bcf0a72b72fd0/torch/distributed/fsdp/fully_sharded_data_parallel.py#L792-L794

I checked in the branch itself with the new unit test too, and I see `istype(subobj, staticmethod)` returning `False`.

Reply:
is it failing due to being wrapped in a contextmanager, or something else?

you may be able to use some utils from `inspect` directly to identify it's a staticmethod.

But my suggestion was also for the Variable part: did you consider using a UserFunctionVariable instead of a GetAttrVariable, as was done for those other staticmethods?

Reply:
> you may be able to use some utils from inspect directly to identify it's a staticmethod.

Unfortunately, I did not see anything relevant in `inspect`.

---

> is it failing due to being wrapped in a contextmanager, or something else?

I think it is something else since for the unit test the method is only a `staticmethod` without being a `contextmanager`.

I think what might be happening is that for `nn.Module`s, we do not use `inspect.getattr_static()`, so we do not bypass the descriptor protocol:
https://github.com/pytorch/pytorch/blob/88c8c2b71b7b2dac0b346dfadc15e96eacaa7755/torch/_dynamo/variables/user_defined.py#L347-L356
As a result, the `isinstance(subobj, staticmethod)` returns `False` unexpectedly.

---

> But my suggestion was also for the Variable part: did you consider using a UserFunctionVariable instead of a GetAttrVariable, as was done for those other staticmethods?

Let me look into this. I used `GetAttrVariable` because FSDP on T5 used to work before https://github.com/pytorch/pytorch/pull/92050, at which point I think the static method would fall through to the final `GetAttrVariable` case:
https://github.com/pytorch/pytorch/blob/88c8c2b71b7b2dac0b346dfadc15e96eacaa7755/torch/_dynamo/variables/user_defined.py#L443
(I was restoring the existing behavior.)

Reply:
For this PR's newly added branch, we pass `subobj` directly as the first arg to `UserFunctionVariable`, while for the existing `staticmethod` branch, we pass `subobj.__get__(self.value)`.

If we pass `subobj.__get__(self.value)` in this newly added branch, we hit something like
```
expected FunctionType found method <bound method TestSingleProc.test_fsdp_staticmethod.<locals>.ModuleWithStaticMethod._add of ModuleWithStaticMethod()>
```
where the first arg's type is incorrect.

Given this, it seems that the newly added branch and the existing `staticmethod` branch have different conditions and different args to `UserFunctionVariable`. I am not sure if we can coalesce them easily.

Reply:
You are right about the `getattr_static` part. If we were not speciailzing for `nn.Modules`, we would have bypassed the descriptor protocol.